unittests/Dialect/ESI/runtime/ESIRuntimeTest.cpp

```diff
@@ -0,0 +1,449 @@
+  EXPECT_EQ(serialized4bit.getSize(), 1UL);
+  EXPECT_EQ(serialized4bit.getData()[0] & 0x0F,
+            0x0F); // Lower 4 bits should be 1111
+
+  auto [deserialized4bit, remaining4] = sint4.deserialize(serialized4bit);
+  auto result4bit = std::any_cast<int64_t>(deserialized4bit);
+  EXPECT_EQ(result4bit, -1);
+
+  // Test maximum negative for 4-bit (-8 = 0x8 in 4 bits)
+  int64_t maxNeg4 = -8;
+  MessageData serializedMaxNeg4 = sint4.serialize(std::any(maxNeg4));
+  auto [deserializedMaxNeg4, remaining5] = sint4.deserialize(serializedMaxNeg4);
+  auto resultMaxNeg4 = std::any_cast<int64_t>(deserializedMaxNeg4);
+  EXPECT_EQ(resultMaxNeg4, -8);
+
+  // Test maximum positive for 4-bit (7 = 0x7 in 4 bits)
+  int64_t maxPos4 = 7;
+  MessageData serializedMaxPos4 = sint4.serialize(std::any(maxPos4));
+  auto [deserializedMaxPos4, remaining6] = sint4.deserialize(serializedMaxPos4);
+  auto resultMaxPos4 = std::any_cast<int64_t>(deserializedMaxPos4);
+  EXPECT_EQ(resultMaxPos4, 7);
+
+  // Test 12-bit signed integers (non-byte-aligned case)
+  SIntType sint12("sint12", 12);
+
+  // Test -1 in 12-bit (should be 0xFFF in lower 12 bits)
+  int64_t minus1w12bit = -1;
+  MessageData serialized12bit = sint12.serialize(std::any(minus1w12bit));
+  EXPECT_EQ(serialized12bit.getSize(), 2UL); // 12 bits = 2 bytes
+
+  auto [deserialized12bit, remaining7] = sint12.deserialize(serialized12bit);
+  auto result12bit = std::any_cast<int64_t>(deserialized12bit);
+  EXPECT_EQ(result12bit, -1);
+
+  // Test a value that requires sign extension: -100 in 12-bit
+  int64_t neg100w12bit = -100;
+  MessageData serializedNeg100 = sint12.serialize(std::any(neg100w12bit));
+  auto [deserializedNeg100, remaining8] = sint12.deserialize(serializedNeg100);
+  auto resultNeg100 = std::any_cast<int64_t>(deserializedNeg100);
+  EXPECT_EQ(resultNeg100, -100);
+}
+
+// Test boundary conditions for sign extension
+TEST(ESITypesTest, SIntTypeSignExtensionBoundaries) {
+  // Test various bit widths to ensure sign extension works correctly
+  for (int width = 1; width <= 16; ++width) {
+    SIntType sintType("sint" + std::to_string(width), width);
+
+    // Calculate the range for this bit width
+    int64_t maxVal = (width == 64) ? INT64_MAX : ((1LL << (width - 1)) - 1);
+    int64_t minVal = (width == 64) ? INT64_MIN : (-(1LL << (width - 1)));
+
+    // Test maximum positive value
+    MessageData serializedMax = sintType.serialize(std::any(maxVal));
+    std::span<const uint8_t> spanMax(serializedMax.getData().data(),
+                                     serializedMax.getSize());
+    auto [deserializedMax, remainingMax] = sintType.deserialize(spanMax);
+    auto resultMax = std::any_cast<int64_t>(deserializedMax);
+    EXPECT_EQ(resultMax, maxVal)
+        << "Failed for width " << width << " max value";
+
+    // Test maximum negative value
+    MessageData serializedMin = sintType.serialize(std::any(minVal));
+    std::span<const uint8_t> spanMin(serializedMin.getData().data(),

```

Comment:
Doesn't MessageData implicitly convert to a span? Why is this necessary? It's not elsewhere.